components/suggest/src/schema.rs

```diff
@@ -193,6 +196,46 @@ CREATE TABLE IF NOT EXISTS dismissed_suggestions (
                 )?;
                 Ok(())
             }
+            19 => {
+                // Clear the database since we're going to be dropping the keywords table and
+                // re-creating it
+                clear_database(tx)?;
+                tx.execute_batch(
+                    "
+-- Recreate the various keywords table to drop the `ON DELETE` clauses
+DROP TABLE keywords;
+DROP TABLE full_keywords;
+DROP TABLE prefix_keywords;
+CREATE TABLE keywords(
+    keyword TEXT NOT NULL,
+    suggestion_id INTEGER NOT NULL REFERENCES suggestions(id),
+    full_keyword_id INTEGER NULL REFERENCES full_keywords(id),
+    rank INTEGER NOT NULL,
+    PRIMARY KEY (keyword, suggestion_id)
+) WITHOUT ROWID;
+CREATE TABLE full_keywords(
+    id INTEGER PRIMARY KEY,
+    suggestion_id INTEGER NOT NULL REFERENCES suggestions(id),
+    full_keyword TEXT NOT NULL
+);
+CREATE TABLE prefix_keywords(
+    keyword_prefix TEXT NOT NULL,
+    keyword_suffix TEXT NOT NULL DEFAULT '',
+    confidence INTEGER NOT NULL DEFAULT 0,
+    rank INTEGER NOT NULL,
+    suggestion_id INTEGER NOT NULL REFERENCES suggestions(id),
+    PRIMARY KEY (keyword_prefix, keyword_suffix, suggestion_id)
+) WITHOUT ROWID;
+CREATE UNIQUE INDEX keywords_suggestion_id_rank ON keywords(suggestion_id, rank);
+-- Create the new indexes
+CREATE INDEX prefix_keywords_suggestion_id ON prefix_keywords(suggestion_id);
+CREATE INDEX keywords_full_keyword_id ON keywords(full_keyword_id);
```

Comment:
What's this used for?

Reply:
It speeds up checking the FK when rows are deleted in the `full_keywords` table.  AFAICT, it has a large impact on performance.  When I delete the index the benchmark times start to skyrocket again.

Reply:
Hmm, interesting.

```
DELETE FROM keywords WHERE suggestion_id IN (SELECT id from suggestions WHERE record_id = :record_id)
DELETE FROM full_keywords WHERE suggestion_id IN (SELECT id from suggestions WHERE record_id = :record_id)
```

The queries don't even use `full_keyword_id` anywhere, not sure how it'd affect the performance. Anyway, not a blocker, just curious.

Reply:
According to the [SQLite docs](https://www.sqlite.org/foreignkeys.html), when a row is deleted when there is a foreign key, it will perform the equivalent of `SELECT rowid FROM <child-table> WHERE <child-key> = :parent_key_value` to check that the constraint is still valid.  I think that's `SELECT rowid FROM keywords WHERE full_keyword_id = :full_keyword_id` for the case of `full_keywords`

Reply:
Ahhh, I see. To enforce the FK invariants.

Now that we don't rely on foreign key cascade delete any more, what if we drop those foreign key constraints from the schema (e.g. drop the FKs on `suggestion_id` and `full_keyword_id` for `keywords`) so it doesn't need to check the FK invariant on deletes. Would that still be sufficiently safe and more performant?

Reply:
I think that's a very reasonable idea.  We lose some safety, but it seems like it shouldn't matter too much in practice -- especially since ingestion in totally deterministic, so if we setup the relationships correctly for one user then it will work for all users.  Performance increases and we save a significant amount of disk space.

I guess my one concern would be the precedent going forward.  I could see us making the same call in a similar situation and it not being correct.  What's your feeling on this?

Reply:
No need to consider that change for this PR, we can continue discussing that offline and make changes later if needed.

Yeah, it's not uncommon to forfeit FK for performance gains in practice. Especially, if the table is mostly write-once-read-many and not storing safety-critical datasets, the safety could be enforced by application itself at the best effort (I think it applies in this case). In the case of things go wrong, we can always drop the whole table and create everything from scratch w/o any user impact. In return, we don't have to pay the FK taxes and it will run faster with lower footprint.

Anyway, we can think of this more offline.